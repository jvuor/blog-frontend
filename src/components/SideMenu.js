import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { Menu, Header } from 'semantic-ui-react'

class SideMenu extends React.Component {

  stickyMenuItem = (blog) =>
    <Menu.Item
      as={Link}
      to={`/${blog.id}/${blog.title}.split(' ').join('-')`}
      key={`${blog.title}-sm`}
      name={`${blog.title}-sm`}
      content={blog.title}
      active={true}
    />

  recentMenuItem = (blog) =>
    <Menu.Item
      as={Link}
      to={`/${blog.id}/${blog.title.split(' ').join('-')}`}
      key={`${blog.title}-rm`}
      name={`${blog.title}-rm`}
      content={blog.title}
      active={true}
    />
  
  render () {
    const stickiedPosts = this.props.blogs.filter(blog => blog.sticky)

    const recentPosts = this.props.blogs.slice(0,5)

    return (
      <Menu vertical>
        <Menu.Item>
          <Header>Jussi Vuorinen</Header>
          <p>Hi, my name is Jussi and I want to be a developer.</p>
        </Menu.Item>      
        {stickiedPosts.length > 0?
          <Menu.Item>
            <Menu.Header>Important posts</Menu.Header>
            <Menu.Menu>
              {stickiedPosts.map(blog => this.stickyMenuItem(blog))}
            </Menu.Menu>
          </Menu.Item>:
          null}
        <Menu.Item>
          <Menu.Header>Recent posts</Menu.Header>
        </Menu.Item>
        <Menu.Menu>
          {recentPosts.map(blog => this.recentMenuItem(blog))}
        </Menu.Menu>
      </Menu>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    blogs: state.blogs
  }
}

export default connect(mapStateToProps, null)(SideMenu)