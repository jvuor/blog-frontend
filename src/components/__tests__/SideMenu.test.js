import React from 'react'
import { SideMenu } from '../SideMenu'
import { MemoryRouter } from 'react-router-dom'
import { shallow } from 'enzyme'
import blogTestData from './blogTestData'

describe('SideMenu component unit tests', () =>{
  test('fails on empty react state', () =>{
    const blogs = null
    const wrapper = shallow(
      <MemoryRouter>
        <SideMenu blogs={blogs} />
      </MemoryRouter>
    )

    expect(() => wrapper.html()).toThrow(TypeError)
  })

  test('snapshot test', () => {
    const blogs = blogTestData
    const wrapper = shallow(
      <MemoryRouter>
        <SideMenu blogs={blogs} />
      </MemoryRouter>
    )
    const content = wrapper.html()

    expect(content).toMatchSnapshot()
  })

  test('displays expected data given a valid state', () => {
    const blogs = blogTestData
    const wrapper = shallow(
      <MemoryRouter>
        <SideMenu blogs={blogs} />
      </MemoryRouter>
    )
    const content = wrapper.html()

    expect(content).toContain('Test Post')
    expect(content).toContain('Important posts')
  })

  test('does not display sticky post menu if there are none', () => {
    const blogs = blogTestData.filter(blog => !blog.sticky)
    const wrapper = shallow(
      <MemoryRouter>
        <SideMenu blogs={blogs} />
      </MemoryRouter>
    )
    const content = wrapper.html()

    expect(content).not.toContain('Important posts')
  })
})