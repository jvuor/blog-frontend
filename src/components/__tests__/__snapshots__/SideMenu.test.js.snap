// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu component unit tests snapshot test 1`] = `"<div class=\\"ui vertical menu\\"><div class=\\"item\\"><div class=\\"ui header\\">Jussi Vuorinen</div><p>Hi, my name is Jussi and I want to be a developer.</p></div><div class=\\"item\\"><div class=\\"header\\">Important posts</div><div class=\\"menu\\"><a class=\\"active item\\" href=\\"/5b55dd6774d88b36481f5607/Second-Test-Post\\">Second Test Post</a></div></div><div class=\\"item\\"><div class=\\"header\\">Recent posts</div></div><div class=\\"menu\\"><a class=\\"active item\\" href=\\"/5b58dce799f2274cc85b0933/Markdown-Test\\">Markdown Test</a><a class=\\"active item\\" href=\\"/5b5895867296874dacbc9d6d/Test-Post-From-Admin-Panel\\">Test Post From Admin Panel</a><a class=\\"active item\\" href=\\"/5b55dd6774d88b36481f5607/Second-Test-Post\\">Second Test Post</a><a class=\\"active item\\" href=\\"/5b55dcdb74d88b36481f5606/Test-post\\">Test post</a></div></div>"`;
